### PR TITLE
Authentication backend for authenticating chef node against chef server

### PR DESCRIPTION
**Pull request by nhuff**
First off this isn't ready to merge yet, but it is far enough along that I would like to get some feedback.

This backend allows a chef node to authenticate itself using its client key.  The basic flow is
1. Client creates a request to login endpoint with mixlib-authentication headers.
2. Vault gets the user id from the request headers and requests the public key from the chef server
3. Vault authenticates the header signature from the request
4. Vault looks up the node information for the client
5. Policies are calculated based on the chef environment, roles, and tags of the client 

The required configuration for the backend is
* base_url: The URL to the chef server's API endpoint
* client_name: The name of the client set up in the chef server for vault to connect with
* client_key: The private key for the client set up in the chef server for vault to connect with

The main issue that I ran into was that HTTP headers aren't really maintained in any of the internal structures.  I had to extend the logical.Connection struct to include the HTTP headers so that I could access them from a logical backend.  This also shows up in the cli implementation that I added.  I have to drop down to using a raw http.Request and http.Client because none of the api.client methods have a way to set request headers. 

I still need to create documentation and tests for the backend, but I wanted get some feedback on whether people thought the general direction was a good one.

### CODE REVIEW

**Comment by jefferai** [comment]
Why not simply pass the header value into Vault as a parameter?

**Comment by nhuff** [comment]
I am not against doing that.  I was trying to maintain the request style that clients currently use to talk to the chef server.  It should be possible to pass the header values in as a json object in the body though. I will change the implementation to do that and see if I run into any other issues.

**Comment by nhuff** [comment]
OK I changed from using headers to including the following in the login request.

* signature_version: This is the value of the X-Ops-Sign header.  I don't use this in the current code, but I intend to support at least versions 1.0, 1.1, and 1.3 of the protocol.
* client_name: Value of the  X-Ops-Userid header.  Client we are authenticating
* timestamp: Value of X-Ops-Timestamp header.
* signature: The base64 encoded signature.  In the Headers this is split over the X-Ops-Authorization-* headers.  Here that value is not split.

The upside is that I don't need the internal changes to pass the Headers around.

The only downside is that the request you are generating the signature for isn't the request that you actually make which is a little weird.  You basically generate the signature for a POST request to the mountpoint/login with an empty body and the actual request is a POST to /v1/mountpoint/login with the above values in the body.  This is required because you would need the signature to generate the signature of the actual request.  Also internally as far as I can tell while the logical backend knows what the mountpoint and path under that it is answering for it doesn't know about anything above that, '/v1' in this case.  This means the backend can't generate the full path to check the request signature against.  This is should be addressable by documentation.

**Comment by jefferai** [comment]
What you're describing does sound weird but I'm also missing why it's necessary. In your description above the path is not one of the items you list as being part of the signature. Or is it?

**Comment by nhuff** [comment]
The signature covers the following attributes of the request.

* Request Path (not host, just path), possibly hashed depending on signature version
* Timestamp
* Client ID
* HTTP Method
* Hash of message body

Version 1.3 of the protocol, which isn't implemented in here yet includes
* Signature Version
* Server API Version

For most things I have seen the server gets the path, message body, and HTTP method from the request itself and uses the time stamp and client id from the headers to validate the signature.  Inside vault I don't have access to the raw request object so I can't get the path, body or method from it.  The method should always be POST so I just hard code that.  What I can get for path is mountpoint/path which is what 'vault write' uses so that probably isn't going to be to odd for anyone using vault.  I also pretty much have to choose an arbitrary known value for the body so I can generate the hashed value which in this case is an empty string.  Before with the version that used the headers the body would have always been an empty string anyway so that isn't really a change.

File: builtin/credential/chef-node/path_config.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+				Type:        framework.TypeString,
+				Description: `Name of the client to connect to chef server with`,
+			},
+			"client_key": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: `PEM encoded client key to use for authenticating to chef server`,
+			},
+		},
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathConfigWrite,
```
Comment (vishalnayak) [inline builtin/credential/chef-node/path_config.go]: Can we add `pathConfigRead` as well?

**Comment by jefferai** [comment]
OK. It seems reasonable to just hardcode '/v1' in front of the known mount path.

File: builtin/credential/chef-node/path_environment.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+		Pattern: "environments/?$",
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ListOperation: b.pathEnvironmentList,
+		},
+	}
+}
+
+func pathEnvironments(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: `environment/(?P<name>.+)`,
```
Comment (vishalnayak) [inline builtin/credential/chef-node/path_environment.go]: Can we move this to `auth/chef-node/config/environment`?

File: builtin/credential/chef-node/path_config.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	"fmt"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func pathConfig(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "config",
```
Comment (vishalnayak) [inline builtin/credential/chef-node/path_config.go]: Can we move this to `auth/chef-node/config/client`?

File: builtin/credential/chef-node/path_tag.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	return &framework.Path{
+		Pattern: "tags/?$",
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ListOperation: b.pathTagList,
+		},
+	}
+}
+
+func pathTags(b *backend) *framework.Path {
+	return &framework.Path{
```
Comment (vishalnayak) [inline builtin/credential/chef-node/path_tag.go]: Can we move this to `auth/chef-node/config/tag`?

File: builtin/credential/chef-node/backend.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package chefNode
```
Comment (vishalnayak) [inline builtin/credential/chef-node/backend.go]: Can we name this backend, just `chef`? This might increase the scope of the backend and will serve as placeholder for enhancements that are out of scope of this PR.

**Comment by nhuff** [comment]
I wasn't sure how stable that /v1 is.  I assume that at some point there might be a /v2.

File: builtin/credential/chef-node/backend.go (hunk 5 above)
Comment (nhuff) [inline builtin/credential/chef-node/backend.go]: The only reason I didn't name it just chef is because you could in theory have a backend that authenticated users against a chef server instead of nodes.  I don't know if anyone would actually want that, but I didn't want to preclude the possibility.

Comment (jefferai) [inline builtin/credential/chef-node/backend.go]: That sounds fine, but make it `chefnode` please. Mixed case package names are uncommon.

**Comment by jefferai** [comment]
There may be but there are no current plans. Unit tests could ensure that if one is created and this isn't updated things break :-D

File: builtin/credential/chef-node/path_config.go (hunk 3 above)
Comment (nhuff) [inline builtin/credential/chef-node/path_config.go]: This could be ambiguous.  While the backend doesn't currently support binding roles to an individual chef client, because I don't think it is a particularly good idea, I could see someone wanting this.  If the other binding endpoints are moved under /config I would think /config/client would be the logical endpoint to use for that.

File: builtin/credential/chef-node/path_environment.go (hunk 2 above)
Comment (jefferai) [inline builtin/credential/chef-node/path_environment.go]: I think it's fine where it is; `environments` is similar to `groups` for LDAP or `roles` for other backends.

Comment (jefferai) [inline builtin/credential/chef-node/path_environment.go]: Or roles for this backend, too :-D

Comment (jefferai) [inline builtin/credential/chef-node/path_environment.go]: I assume the final policy values will be combined from environment and/or tag and/or role?

File: builtin/credential/chef-node/path_config.go (hunk 3 above)
Comment (nhuff) [inline builtin/credential/chef-node/path_config.go]: I suppose /config/node would also make sense for the per node policy bindings.  And I see /config/client has precedent in the AWS backend so I will go ahead and move under /config/client.

File: builtin/credential/chef-node/path_environment.go (hunk 2 above)
Comment (nhuff) [inline builtin/credential/chef-node/path_environment.go]: Yes you will get the union of the roles.  There didn't really seem to be any standard in the existing backends for how these should be named.  I don't have a strong preference either way.

File: builtin/credential/chef-node/path_config.go (hunk 3 above)
Comment (jefferai) [inline builtin/credential/chef-node/path_config.go]: In the code I only see role/tag/environments; what would config/node (or just node) be for? If it's per-node policy bindings it should stay out of config/.

**Comment by jefferai** [comment]
@nhuff It seems like this is mostly implemented but not a unit test in sight. What's the right way to go about getting this tested? Ideally there'd be a Docker container that could be populated as part of the test setup.

**Comment by nhuff** [comment]
Where are we on the /config endpoint?  Right now I haven't moved anything so it is still
/config
/environment
/tag
/role
Is that what you want to go with?

Tests are the next thing on my list.  I have been manually testing against an internal Chef enterprise server while developing so I need to figure out the best way to do the automated integration tests.

The other thing I would consider a blocker for integration is documentation.  After I figure out the tests I need to write that.

**Comment by jefferai** [comment]
From my perspective I was waiting on an answer to what tags, environments, roles, and nodes are. If they are Chef constructs that you are matching against, I think they should be where they are now. If they are part of the client configuration, they should go under config, e.g. with the other client config at config/client.

(I know the answer is likely the former, just being explicit.)

**Comment by nhuff** [comment]
Yes they are Chef constructs.  A node is an individual Chef client, usually a server.  Nodes are in an environment and can have tags and roles associated with them.  The code doesn't currently support directly binding policy to individual nodes right now, but I could see that being a future extension someone would want.

**Comment by nhuff** [comment]
What is the preferred way to do unit testing?  Looking at the current auth backends some of them use the github.com/hashicorp/vault/logical/testing package, some just use golang's testing directly, and some use both with no rhyme or reason as to which one they are using that I can figure out.

**Comment by jefferai** [comment]
Hi @nhuff ,

The logical/testing package is what most of the early backend tests were written with, but it's fallen out of favor with the current dev team because there is purpose to its dogmatic approach (e.g. ensuring proper separation of actual interfaces) but it didn't really anticipate how complicated some of the backends could be and as a result it can be extremely difficult to actually write many types of tests without a huge amount of ugliness. Look at the PKI backend if you want some eye bleeding.

So, most older tests use that framework; most newer tests don't. If you see a mix, likely the ones not using that framework are newer tests in the same package.

The approach we have been using with third party services has been using `dockertest` as you can see in e.g. postgres. It'd be great if the tests started a real chef server, populated with nodes as appropriate, and then ran against that. There is a v3 API for `dockertest` that is much more generic and we'd like to be targeting but doesn't allow advanced container options currently, e.g. bind mounting, so v2 might be necessary.

**Comment by jefferai** [comment]
At this point I'm going to slot in for 0.7 (probably early March) but we remain very interested in it so will keep it going with you!

**Comment by nhuff** [comment]
No problem, I thought 0.6.5 was kind of ambitious anyway. My current plan is to work on unit tests, then documentation, then the acceptance test.  I might try and start with acceptance tests that run against Hosted Chef and then move over to a docker image.  The tests should transfer pretty well from Hosted Chef to a docker image once I get time to create a docker image.

**Comment by jefferai** [comment]
Huh, I naively assumed that a chef server would be part of the docker library.

**Comment by nhuff** [comment]
I found one for a very old version of chef server, but nothing current.

**Comment by nhuff** [comment]
I have tests and documentation added.  There is a test failure, but it doesn't appear to be in anything I touched.

I also added a /client endpoint that allows binding policy to specific Chef clients.

**Comment by nhuff** [comment]
Merging with latest master has fixed the failing test.

**Comment by coderanger** [comment]
Just from a quick read over of this, any use of this backend would be hilariously unsafe as it is basing which policies to assign to the node on things the node can change itself.

**Comment by luckymike** [comment]
To be clear, @coderanger is referring to:
>Policies are calculated based on the chef environment, roles, and tags of the client

**Comment by coderanger** [comment]
https://coderanger.net/chef-and-vault/ goes in to more detail on the problem.

**Comment by nhuff** [comment]
That is why I added a way to bind policy to the client key and put warnings in the documentation about using the other methods.

**Comment by mfischer-zd** [comment]
@nhuff Good point, but is anyone actually going to use the client key as a binding, especially in large installations?  That seems like an awfully inconvenient user experience.

**Comment by coderanger** [comment]
I mean its up to Vault as a project but supporting anything other than per-client policies is literally just asking the client what policy it wants but with a few more steps.

File: builtin/credential/chef-node/path_login.go [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+	var jb nodeResponse
+	if err := json.Unmarshal(body, &jb); err != nil {
+		return nil, err
+	}
+
+	return &jb, nil
+}
+
+func (b *backend) retrievePubKey(req *logical.Request, targetName string) (*rsa.PublicKey, error) {
```
Comment (coderanger) [inline builtin/credential/chef-node/path_login.go]: This will only retrieve the first key for a client, you need to use the https://docs.chef.io/api_chef_server.html#clients-client-keys API to see other keys.

Comment (nhuff) [inline builtin/credential/chef-node/path_login.go]: How do you get more than one key? I saw that API, but couldn't find any nodes in our system that had more than one.

Comment (coderanger) [inline builtin/credential/chef-node/path_login.go]: Via https://docs.chef.io/knife_client.html#key-create and similar commands. We still treat the "primary" key as special for backwards compat but you can create as many as you need. This mostly exists for key rotation.

**Comment by jefferai** [comment]
There was definitely some misunderstanding on our end. I for one had no idea chef nodes could update all of their own info (and, why is that a good idea even), I thought it was more like Ansible or Salt where info about a client comes from a central store.

We definitely can't support an authentication method where the client can choose its own policies.

**Comment by coderanger** [comment]
@jefferai I wouldn't call it so much a "good idea" as "very old tech debt". We have a plan to fix it, but we've been saying that for a while now (years) and even with that the best guidance I could give would be "within 6 months hopefully".

The per-client mappings are still somewhat useful if you've got an external non-Chef-Server source of truth for role mappings (like a git repo with trusted Chef-y data) you could use scripts to sync that with Vault but that is going to be a lot of work and I'm afraid I have to agree with @mfischer-zd that very few people would want to use this.

That said, once we get Chef Server fixed up this goes from Very Bad Idea (on the role/env/tag mappings) to Oh Yes Please Right Now so do with this knowledge what you will.

**Comment by nhuff** [comment]
While binding to clients individually is more manually intensive, this is how chef-vault works now so it isn't like people aren't already dealing with it in other systems. 

In my case I have probably a couple dozen hosts that need access to highly secure credentials where I would probably just setup the mappings by hand. 

I have some other things where it would be nice to use vault to generate TLS certs that identify a server and don't grant access to data. I can also audit if a cert gets used on more than one system so I would be fine mapping to an environment for this use case.

**Comment by jefferai** [comment]
@coderanger Thanks for the further color.

This puts me in a bit of a bind because I know @nhuff worked hard on this; at the same time had I understood this detail I'd have put a stop to it a while back.

I think the best thing to do would be to table this until Chef implements the changes you described, at which point it could be updated to work best with the new paradigms. Does that make sense?

**Comment by coderanger** [comment]
@nhuff The difference with chef-vault is that has the `knife vault refresh` process to re-run the encryption stuffs, and being a manual process you (a human) could hopefully check if someone had done something evil. With Vault being an online system, there would be no human to notice the badness. Also this is super bad in chef-vault too and makes me wish people didn't use it :-(

**Comment by coderanger** [comment]
And sorry again if I came off as a buzzkill, I reeeeeally want this feature, just also don't want to give people more footguns given how many people misunderstand the risk factors.

**Comment by nhuff** [comment]
So client bindings have a more involved workflow, but are secure. I can remove the other endpoints in a commit and if chef ever fixes the server side we could revert that commit.

**Comment by mfischer-zd** [comment]
I'd very much like to see the inverse workflow, where Vault becomes a trusted third party to Chef Server.  That way a new node can get "blessed" by Vault signing its certificate request, and Chef Server will trust Vault's CA.

**Comment by thommay** [comment]
The work @coderanger is describing is set out in https://chef.github.io/chef-rfc/rfc045-node_state_separation.html, more or less. It's likely (but not completely confirmed) that that work will happen this year.

**Comment by nhuff** [comment]
OK so this now only supports binding policies to individual clients and supports multiple client keys.

**Comment by Happycoil** [comment]
Any estimate on when we can expect this? Even the client binding-only version would be _great_.

**Comment by directionless** [comment]
To echo @Happycoil, binding to clients sounds pretty good to me. I think the would alleviate a bunch of the security concerns, and I think it provides a really powerful way to bootstrap.

**Comment by jefferai** [comment]
I think before reviewing it would be nice if @coderanger and/or @thommay can comment on whether this is a good idea vs. just waiting until the necessary bits are in Chef server. Rationale is the comment:

> The per-client mappings are still somewhat useful if you've got an external non-Chef-Server source of truth for role mappings (like a git repo with trusted Chef-y data) you could use scripts to sync that with Vault but that is going to be a lot of work and I'm afraid I have to agree with @mfischer-zd that very few people would want to use this.

I realize that some of those people are here on the ticket, but one of my concerns is introducing a backend that works for this narrow use-case but requires a lot of deprecating/changing/rearchitecting once the more ideal solution comes out. We've been down that road before and it's really unfun for both devs and users.

**Comment by nhuff** [comment]
I actually don't think it should require any deprecation.  There may be nicer ways to set up the policy mappings once chef fixes the server side issues, but I don't see any reason to deprecate binding to a specific client.

**Comment by nhuff** [comment]
I have pulled this out into an external plugin here https://github.com/nhuff/vault-plugin-auth-chefnode.  It currently works and I will be adding documentation shortly.

**Comment by jefferai** [comment]
I would suggest the documentation include appropriate caveats around the dangers that currently exist until Chef implements the server side enhancements.

Closing for now; if you want it to be included as a built in, please submit a separate PR for that, although I think we'll want to wait for the objections currently raised to be addressed before doing so.

Thanks!

**Comment by nhuff** [comment]
As far as I know all of the objections raised were addressed by limiting binding only to individual clients.  I don't really care whether it lives in or out of tree.  It just seemed totally stalled here.

**Comment by jefferai** [comment]
@nhuff I don't think anyone on our end knew this was waiting for a review.
